src/v/model/record_batch_reader.cc

```diff
@@ -215,7 +215,7 @@ make_fragmented_memory_storage_batches(Container batches) {
     size_t i = 0;
     for (auto it = batches.begin(); it != batches.end(); ++i, ++it) {
         if (!data_chunk.empty() && i % elements_per_fragment == 0) {
-            if (is_foreign) {
+            if constexpr (is_foreign) {
```

Comment:
Does it make any sense to use constexpr here? inside the branch there is nothing that depends on `is_foreign` value. The compiler should be able to eliminate the dead branch if `is_foreign` is false.

Reply:
Likely this will produce the same code as a non-constexpr `if`, but I like to see `if constexpr` when evaluating compile time constants (`is_foreign` is a non-type template argument) just for completeness sake, and the confidence that the dead branch will in fact be removed.